docs/API_DESIGN.md

```diff
@@ -0,0 +1,658 @@
+
+  /// The metadata associated with the DID resolution process.
+  PUBLIC DATA resolution_metadata: ResolutionMetadata
+
+  /// Resolve via a DID URI.
+  CONSTRUCTOR resolve(uri: string)
+```
+
+**Examples**
+
+```pseudocode!
+🚧
+```
+
+### `ResolutionMetadataError`
+
+```pseudocode!
+/// The error code from the resolution process.
+ENUM ResolutionMetadataError
+  invalidDid
+  notFound
+  representationNotSupported
+  methodNotSupported
+  invalidDidDocument
+  invalidDidDocumentLength
+  internalError
+```
+
+### `ResolutionMetadata`
+
+```pseudocode!
+/// Metadata about the given resolution.
+CLASS ResolutionMetadata
+  /// The error code from the resolution process.
+  PUBLIC DATA error: ResolutionMetadataError?
+```
+
+### `DocumentMetadata`
+
+```pseudocode!
+/// Metadata about the DID Document.
+CLASS DocumentMetadata
+  /// 🚧
+  PUBLIC DATA created: string?
+
+  /// 🚧
+  PUBLIC DATA updated: string?
+
+  /// 🚧
+  PUBLIC DATA deactivated: bool?
+
+  /// 🚧
+  PUBLIC DATA nextUpdate: string?
+
+  /// 🚧
+  PUBLIC DATA versionId: string?
+
+  /// 🚧
+  PUBLIC DATA nextVersionId: string?
+
+  /// 🚧
+  PUBLIC DATA equivalentId: []string?
+
+  /// 🚧
+  PUBLIC DATA canonicalId: string?
+```
+
+## Methods
+
+### `DidJwk`
+
+```pseudocode!
+CLASS DidJwk
+  PUBLIC DATA did: Did
+  PUBLIC DATA document: Document
+  CONSTRUCTOR(public_key: Jwk)
+  CONSTRUCTOR(uri: string)
+  STATIC METHOD resolve(uri: string): ResolutionResult
+```
+
+#### Example: Create a `did:jwk`
+
+```pseudocode!
+key_manager = new InMemoryKeyManager()
+public_key = key_manager.generate_key_material()
+did_jwk = new DidJwk(public_key)
+```
+
+#### Example: Instantiate an existing `did:jwk`
+
+```pseudocode!
+uri = "did:jwk:eyJrdHkiOiJFQyIsInVzZSI6InNpZyIsImNydiI6InNlY3AyNTZrMSIsImtpZCI6ImkzU1BSQnRKS292SEZzQmFxTTkydGk2eFFDSkxYM0U3WUNld2lIVjJDU2ciLCJ4IjoidmRyYnoyRU96dmJMRFZfLWtMNGVKdDdWS04VEZaTm1BOVlnV3p2aGg3VSIsInkiOiJWTEZxUU1aUF9Bc3B1Y1hvV1gyLWJHWHBBTzFmUTVMbjE5VjVSQXhyZZVIiwiYWxnIjoiRVMyNTZLIn0"
+did_jwk = new DidJwk(uri)
+```
+
+#### Example: Resolve a `did:jwk`
+
+```pseudocode!
+uri = "did:jwk:eyJrdHkiOiJFQyIsInVzZSI6InNpZyIsImNydiI6InNlY3AyNTZrMSIsImtpZCI6ImkzU1BSQnRKS292SEZzQmFxTTkydGk2eFFDSkxYM0U3WUNld2lIVjJDU2ciLCJ4IjoidmRyYnoyRU96dmJMRFZfLWtMNGVKdDdWS04VEZaTm1BOVlnV3p2aGg3VSIsInkiOiJWTEZxUU1aUF9Bc3B1Y1hvV1gyLWJHWHBBTzFmUTVMbjE5VjVSQXhyZZVIiwiYWxnIjoiRVMyNTZLIn0"
+resolution_result = DidJwk.resolve(uri)
+```
+
+### `DidWeb`
+
+```pseudocode!
+CLASS DidWeb
+  CONSTRUCTOR(uri: string)
+  STATIC METHOD resolve(uri: string): ResolutionResult
+```
+
+#### Example: Instantiate an existing `did:web`
+
+```pseudocode!
+uri = "did:web:example.com"
+did_web = new DidWeb(uri)
+```
+
+#### Example: Resolve a `did:web`
+
+```pseudocode!
+uri = "did:web:example.com"
+resolution_result = DidWeb.resolve(uri)
+```
+
+### `DidDht`
+
+```pseudocode!
+CLASS DidDht
+  PUBLIC DATA did: Did
+  PUBLIC DATA document: Document
+  CONSTRUCTOR(identity_key: Jwk)
+  CONSTRUCTOR(uri: string)
```

Comment:
how could i create a DidDht with knowing just the uri?

Reply:
central to what i want to ask here is 

> what can i _do_ with a DidWhateverMethod once instantiated?

Reply:
The `DidDht` is an encapsulation around the `Did` and `Document` abstraction, both of those are accessible via the DID URI. Admittedly I'm not sure this has a concrete use case right now 🤔

Reply:
> in my mind, DIDs (perhaps it's more BearerDids) serve one purpose - to be a vessel that identifies me and holds a signing key so i can prove to bob that a thing i sent him actually came from me

Yeah you're referring specifically to bearer DID's, which is our own concept. The concept of a DID is actually moreso on the matter of accessing public keys, so for example if one wanted to verify a cryptographic signature and they had a DID URI, they could call this to gain access to the public key for verification.